app/controllers/concerns/compression.rb

```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module Compression
+  extend ActiveSupport::Concern
+
+  def deflate(data, **) = Zlib::Deflate.deflate(data, **)
+  def gzip(data, deterministic: true, **)
+    zipped = StringIO.new
+    zipped.set_encoding(Encoding::BINARY)
+
+    gz = Zlib::GzipWriter.new(zipped, Zlib::BEST_COMPRESSION)
+    gz.mtime = 0 if deterministic
```

Comment:
Gzip has an internal timestamp by default, which makes it non-deterministic, which breaks etags.